src/tsd.c

```diff
@@ -302,7 +302,21 @@ tsd_fetch_slow(tsd_t *tsd, bool minimal) {
 			tsd_data_init_nocleanup(tsd);
 		}
 	} else if (tsd_state_get(tsd) == tsd_state_minimal_initialized) {
-		if (!minimal) {
+		/*
+		 * If a thread only ever deallocates (e.g. dedicated reclamation
+		 * threads), we want to help it to eventually escape the slow
+		 * path (caused by the minimal initialized state).  Since the
+		 * total deallocated bytes is maintained all the time, it can
+		 * tell if plenty of deallocation activities already happened,
+		 * in which case the tsd is switched to nominal.
+		 *
+		 * The 8K threshold means at most 1K of dealloc calls stay on
+		 * the slow path.
+		 *
+		 * Also see comments in free_default().
+		 */
+		if (!minimal || (*tsd_thread_deallocatedp_get(tsd) >=
+		    TSD_STATE_MINIMAL_INITIALIZED_DEALLOC_THRESHOLD)) {
```

Comment:
@davidtgoldblatt : this is the workaround for the dealloc-only threads slow TSD issue we discussed, in case you want a reference.